src/closest.ts

```diff
@@ -0,0 +1,170 @@
+  flatCoordinates: number[],
+  offset1: number,
+  offset2: number,
+  stride: number,
+  x: number,
+  y: number,
+  closestPoint: number[]
+) {
+  const x1 = flatCoordinates[offset1];
+  const y1 = flatCoordinates[offset1 + 1];
+  const dx = flatCoordinates[offset2] - x1;
+  const dy = flatCoordinates[offset2 + 1] - y1;
+  let offset;
+  if (dx === 0 && dy === 0) {
+    offset = offset1;
+  } else {
+    const t = ((x - x1) * dx + (y - y1) * dy) / (dx * dx + dy * dy);
+    if (t > 1) {
+      offset = offset2;
+    } else if (t > 0) {
+      for (let i = 0; i < stride; ++i) {
+        closestPoint[i] = lerp(
+          flatCoordinates[offset1 + i],
+          flatCoordinates[offset2 + i],
+          t
+        );
+      }
+      closestPoint.length = stride;
+      return;
+    } else {
+      offset = offset1;
+    }
+  }
+  for (let i = 0; i < stride; ++i) {
+    closestPoint[i] = flatCoordinates[offset + i];
+  }
+  closestPoint.length = stride;
+}
+
+type ClosestPointResult = {
+  totalSquaredDistance: number;
+  squaredDistance: number;
+  closestPoint: number[];
+}
+
+export function getClosestPoint(
+  flatCoordinates: number[],
+  offset: number,
+  end: number,
+  stride: number,
+  maxDelta: number,
+  x: number,
+  y: number
+): ClosestPointResult {
+  let minSquaredDistance = Infinity;
+  const closestPoint = [NaN, NaN];
+  let minIndex;
+  if (offset == end) {
+    return {
+      closestPoint: closestPoint,
+      squaredDistance: minSquaredDistance,
+      totalSquaredDistance: minSquaredDistance,
+    };
+  }
+  let i, squaredDistance;
+  if (maxDelta === 0) {
+    // All points are identical, so just test the first point.
+    squaredDistance = squaredDx(
+      x,
+      y,
+      flatCoordinates[offset],
+      flatCoordinates[offset + 1]
+    );
+    if (squaredDistance < minSquaredDistance) {
+      for (i = 0; i < stride; ++i) {
+        closestPoint[i] = flatCoordinates[offset + i];
+      }
+      closestPoint.length = stride;
+      return {
+        closestPoint: closestPoint,
+        squaredDistance: squaredDistance,
+        totalSquaredDistance: squaredDx(
+          flatCoordinates[offset],
+          flatCoordinates[offset + 1],
+          closestPoint[0],
+          closestPoint[1]
+        ),
+      };
+    }
+    return {
+      closestPoint: closestPoint,
+      squaredDistance: minSquaredDistance,
+      totalSquaredDistance: minSquaredDistance,
+    };
+  }
+  const tmpPoint = [NaN, NaN];
+  let index = offset + stride;
+  while (index < end) {
+    assignClosest(
+      flatCoordinates,
+      index - stride,
+      index,
+      stride,
+      x,
+      y,
+      tmpPoint
+    );
+    squaredDistance = squaredDx(x, y, tmpPoint[0], tmpPoint[1]);
+    if (squaredDistance < minSquaredDistance) {
+      minSquaredDistance = squaredDistance;
+      minIndex = index;
+      for (i = 0; i < stride; ++i) {
+        closestPoint[i] = tmpPoint[i];
+      }
+      closestPoint.length = stride;
```

Comment:
I think we could avoid these copies by defining 2 points and swapping between them appropriately:
```js
let closestPoint = [];
let assignedPoint = [];
...
assignedPoint = assignClosest(..., assignedClosest);
if (squaredDistance < minSquaredDistance) {
   minSquaredDistance = squaredDistance;
   minIndex = index;
  [closestPoint, assignedPoint] = [assignedPoint, closestPoint]; // swap variables
...
}
```